File: shared/aaaaaa Accuracy ad key files/google_business.js
// google_business.js

document.addEventListener("DOMContentLoaded", () => {
    const accountsContainer = document.getElementById("accounts-container");
    const locationsContainer = document.getElementById("locations-container");
    const reviewsChart = document.getElementById("reviews-chart");
    const keywordsList = document.getElementById("keywords-list");
    const generateReportBtn = document.getElementById("generate-report-btn");
    const locationModal = document.getElementById("location-modal");
    const closeModalBtn = document.getElementById("close-modal-btn");
    const locationDetails = document.getElementById("location-details");

    let chartInstance = null;

    /**
     * Fetch Google Business Accounts from the Backend.
     */
    async function fetchAccounts() {
        try {
            const response = await fetch("/api/google_business/accounts");
            const data = await response.json();

            if (data.accounts && data.accounts.length > 0) {
                accountsContainer.innerHTML = data.accounts
                    .map(
                        (account) =>
                            `<div class="account-item" data-account-id="${account.id}">
                                <h3>${account.name}</h3>
                                <p>${account.email}</p>
                                <button class="btn-view-locations" data-account-id="${account.id}">View Locations</button>
                            </div>`
                    )
                    .join("");
                attachAccountListeners();
            } else {
                accountsContainer.innerHTML = "<p>No accounts found.</p>";
            }
        } catch (error) {
            accountsContainer.innerHTML = `<p>Error fetching accounts: ${error.message}</p>`;
        }
    }

    /**
     * Attach Event Listeners to Account Buttons.
     */
    function attachAccountListeners() {
        const locationButtons = document.querySelectorAll(".btn-view-locations");
        locationButtons.forEach((button) => {
            button.addEventListener("click", () => {
                const accountId = button.dataset.accountId;
                fetchLocations(accountId);
            });
        });
    }

    /**
     * Fetch Locations for a Specific Account.
     */
    async function fetchLocations(accountId) {
        try {
            const response = await fetch(`/api/google_business/accounts/${accountId}/locations`);
            const data = await response.json();

            if (data.locations && data.locations.length > 0) {
                locationsContainer.innerHTML = data.locations
                    .map(
                        (location) =>
                            `<div class="location-item" data-location-id="${location.id}">
                                <h4>${location.name}</h4>
                                <p>${location.address}</p>
                                <button class="btn-view-details" data-location-id="${location.id}">View Details</button>
                            </div>`
                    )
                    .join("");
                attachLocationListeners();
            } else {
                locationsContainer.innerHTML = "<p>No locations found.</p>";
            }
        } catch (error) {
            locationsContainer.innerHTML = `<p>Error fetching locations: ${error.message}</p>`;
        }
    }

    /**
     * Attach Event Listeners to Location Buttons.
     */
    function attachLocationListeners() {
        const detailButtons = document.querySelectorAll(".btn-view-details");
        detailButtons.forEach((button) => {
            button.addEventListener("click", () => {
                const locationId = button.dataset.locationId;
                fetchLocationDetails(locationId);
            });
        });
    }

    /**
     * Fetch and Display Location Details in Modal.
     */
    async function fetchLocationDetails(locationId) {
        try {
            const response = await fetch(`/api/google_business/locations/${locationId}`);
            const data = await response.json();

            locationDetails.innerHTML = `
                <h3>${data.name}</h3>
                <p>${data.address}</p>
                <p>Phone: ${data.phone}</p>
                <p>Status: ${data.status}</p>
                <p>Ratings: ${data.ratings.average} (${data.ratings.total} reviews)</p>
            `;
            locationModal.classList.remove("hidden");
        } catch (error) {
            locationDetails.innerHTML = `<p>Error fetching location details: ${error.message}</p>`;
        }
    }

    /**
     * Close Location Modal.
     */
    closeModalBtn.addEventListener("click", () => {
        locationModal.classList.add("hidden");
    });

    /**
     * Fetch and Render Reviews Analysis.
     */
    async function fetchReviewsAnalysis() {
        try {
            const response = await fetch("/api/google_business/reviews");
            const data = await response.json();

            if (chartInstance) chartInstance.destroy(); // Clear existing chart

            chartInstance = new Chart(reviewsChart, {
                type: "pie",
                data: {
                    labels: ["Positive", "Neutral", "Negative"],
                    datasets: [
                        {
                            data: [data.positive, data.neutral, data.negative],
                            backgroundColor: ["#4CAF50", "#FFC107", "#F44336"],
                        },
                    ],
                },
            });

            keywordsList.innerHTML = data.keywords
                .map((keyword) => `<li>${keyword}</li>`)
                .join("");
        } catch (error) {
            keywordsList.innerHTML = `<p>Error fetching reviews analysis: ${error.message}</p>`;
        }
    }

    /**
     * Generate a Report.
     */
    generateReportBtn.addEventListener("click", async () => {
        try {
            const response = await fetch("/api/google_business/report", {
                method: "POST",
            });
            const data = await response.json();

            if (data.report_url) {
                alert("Report generated successfully!");
                window.open(data.report_url, "_blank");
            } else {
                alert("Failed to generate report.");
            }
        } catch (error) {
            alert(`Error generating report: ${error.message}`);
        }
    });

    // Initial Load
    fetchAccounts();
    fetchReviewsAnalysis();
});

File: shared/aaaaaa Accuracy ad key files/google_business.py
import requests
import logging
from typing import Dict, List, Any

class GoogleBusinessManager:
    """
    A tool to manage and optimize Google Business Profiles for RLG Data and RLG Fans clients.
    Provides functionalities for fetching business information, updating profiles, and analyzing reviews.
    """

    def __init__(self, api_key: str):
        """
        Initialize the GoogleBusinessManager with the API key.
        """
        self.api_key = api_key
        self.base_url = "https://mybusiness.googleapis.com/v4"

        # Set up logging
        logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")

    def get_locations(self) -> List[Dict[str, Any]]:
        """
        Retrieve the list of locations associated with the Google Business account.
        """
        url = f"{self.base_url}/accounts"
        params = {"key": self.api_key}

        response = requests.get(url, params=params)
        if response.status_code == 200:
            accounts = response.json().get("accounts", [])
            locations = []
            for account in accounts:
                account_id = account.get("name", "").split("/")[-1]
                locations_url = f"{self.base_url}/accounts/{account_id}/locations"
                locations_response = requests.get(locations_url, params=params)
                if locations_response.status_code == 200:
                    locations.extend(locations_response.json().get("locations", []))
                else:
                    logging.error(f"Failed to fetch locations for account {account_id}: {locations_response.text}")
            return locations
        else:
            logging.error(f"Failed to fetch accounts: {response.text}")
            return []

    def update_location_details(self, location_id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
        """
        Update details for a specific location.
        """
        url = f"{self.base_url}/accounts/{location_id}/locations"
        params = {"key": self.api_key}
        response = requests.patch(url, json=updates, params=params)
        
        if response.status_code == 200:
            logging.info(f"Location {location_id} updated successfully.")
            return response.json()
        else:
            logging.error(f"Failed to update location {location_id}: {response.text}")
            return {"error": response.text}

    def fetch_reviews(self, location_id: str) -> List[Dict[str, Any]]:
        """
        Fetch reviews for a specific location.
        """
        url = f"{self.base_url}/accounts/{location_id}/locations/reviews"
        params = {"key": self.api_key}

        response = requests.get(url, params=params)
        if response.status_code == 200:
            return response.json().get("reviews", [])
        else:
            logging.error(f"Failed to fetch reviews for location {location_id}: {response.text}")
            return []

    def analyze_reviews(self, reviews: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Analyze customer reviews for sentiment and recurring themes.
        """
        positive, negative, neutral = 0, 0, 0
        keywords = {}

        for review in reviews:
            comment = review.get("comment", "").lower()
            rating = review.get("starRating", "")

            if rating in ["POSITIVE", "FIVE_STAR"]:
                positive += 1
            elif rating in ["NEGATIVE", "ONE_STAR"]:
                negative += 1
            else:
                neutral += 1

            for word in comment.split():
                keywords[word] = keywords.get(word, 0) + 1

        return {
            "total_reviews": len(reviews),
            "positive": positive,
            "negative": negative,
            "neutral": neutral,
            "top_keywords": sorted(keywords.items(), key=lambda x: x[1], reverse=True)[:10]
        }

    def generate_report(self, location_id: str) -> Dict[str, Any]:
        """
        Generate a detailed report for a Google Business Profile location.
        """
        reviews = self.fetch_reviews(location_id)
        analysis = self.analyze_reviews(reviews)
        location_details = self.get_location_details(location_id)

        return {
            "location_details": location_details,
            "review_analysis": analysis,
            "total_reviews": analysis.get("total_reviews"),
        }

    def get_location_details(self, location_id: str) -> Dict[str, Any]:
        """
        Fetch detailed information about a specific location.
        """
        url = f"{self.base_url}/accounts/{location_id}/locations"
        params = {"key": self.api_key}
        response = requests.get(url, params=params)
        
        if response.status_code == 200:
            return response.json()
        else:
            logging.error(f"Failed to fetch details for location {location_id}: {response.text}")
            return {"error": response.text}


if __name__ == "__main__":
    # Replace with a valid API key
    api_key = "YOUR_GOOGLE_BUSINESS_API_KEY"

    google_business = GoogleBusinessManager(api_key=api_key)
    print("Fetching all business locations...")
    locations = google_business.get_locations()

    if locations:
        print(f"Found {len(locations)} locations. Generating reports...")
        for location in locations:
            location_id = location.get("name", "").split("/")[-1]
            report = google_business.generate_report(location_id)
            print(f"Report for location {location_id}:")
            print(report)
